panel/widgets/remote_file_selector.py

```diff
@@ -0,0 +1,379 @@
+import asyncio
+import threading
+
+from abc import abstractmethod
+from pathlib import PurePosixPath
+from typing import (
+    ClassVar, Dict, List, Optional, Type,
+)
+
+import param
+
+from panel.layout import (
+    Column, Divider, ListPanel, Row,
+)
+
+from ..pane.markup import Markdown
+from ..viewable import Layoutable
+from .base import CompositeWidget
+from .button import Button
+from .input import TextInput
+from .select import CrossSelector
+
+
+class RemoteFileProvider(param.Parameterized):
+    """
+    `RemoteFileProvider` is an abstract class used by `RemoteFileSelector` to
+    list content of a remote filesystem.
+
+    Reference: https://panel.holoviz.org/reference/widgets/RemoteFileSelector.html
+
+    """
+    def __init__(self):
+        super().__init__()
+
+    @abstractmethod
+    def ls(self, path):
+        """
+        Concrete classes must implement this method to list the content of a remote filesystem.
+
+        Arguments
+        ---------
+        path: str
+            The path to search
+
+        Returns
+        -------
+        A tuple of two lists: the first one contains the directories, the second one contains the files.
+        Each element of the lists is a string representing the *name* (not the full path) of the directory or file.
+
+
+        """
+        raise NotImplementedError()
+
+# for S3RemoteFileProvider
+import s3fs
```

Comment:
There is no need to make this a required dependency. 

As far as I can, see this can be put into a `if TYPE_CHECKING`